### PR TITLE
Introduce option to hide dashboard pages.

### PR DESCRIPTION
**Pull request by lasanthaS**
## Purpose
This PR introduces an option in the dashboard designer pages panel to hide specific pages. 

Resolves https://github.com/wso2/carbon-dashboards/issues/983

### CODE REVIEW

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 1]
```diff
@@ -197,14 +197,35 @@ class DashboardViewPage extends Component {
...
+
+    /**
+     * Render pages menu.
+     *
+     * @returns {XML} Html content
+     */
     renderPagesList() {
         const landingPage = this.dashboard.landingPage;
         const history = this.props.history;
         return this.dashboard.pages.map((page) => {
```
Comment (this) [inline components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: With the new `hidden`, all `page` objects in the `pages` array won't produce `<ListItem>` element. Hence, instead of `Array.map()` lets use `Array.forEach()` here. Then you won't be needed to `return '';`.

Comment (this) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]:
```js
const listItems = [];
return this.dashboard.pages.forEach((page) => {
     if (this.isPageHidden(page)) {
          return;
     }
     // create listItem = <ListItem>
     listItems.push(listItem);
}
```

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx [hunk 2]
```diff
@@ -170,6 +179,37 @@ class PagesSidePane extends Component {
...
+            const page = this.getPageById(pageId);
+            if (!page) {
+                this.setValidationErrorMessage(`Cannot find page for ID '${pageId}'!`);
+                reject();
+            }
+            page.hidden = hidden;
+            this.props.updateDashboard()
+                .then(() => resolve())
+                .catch(() => {
+                    delete page.title;
```
[the remark this answers is not in the document]
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: It should be `delete page.hidden`. Thanks for finding that out!

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx [hunk 3]
```diff
@@ -170,6 +179,37 @@ class PagesSidePane extends Component {
...
+     * Update dashboard page visibility.
+     *
+     * @param {string} pageId Page ID
+     * @param {boolean} hidden Is hidden
+     * @returns {Promise}
+     */
+    handlePageVisibility(pageId, hidden) {
+        return new Promise((resolve, reject) => {
+            const page = this.getPageById(pageId);
+            if (!page) {
```
[the remark this answers is not in the document]
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: Basically not

File: components/dashboards-web-component/src/designer/components/PageCard.jsx [hunk 4]
```diff
@@ -151,6 +151,18 @@ class PageCard extends Component {
                                 disabled={page.id === this.props.landingPageId}
                                 onCheck={() => this.props.updateLandingPage(this.props.page.id)}
                             />
+                            {
+                                /* Show hide page option only for non-landing page pages. */
+                                page.id !== this.props.landingPageId &&
```
[the remark this answers is not in the document]
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PageCard.jsx]: Yes, this is where we do that logic in the UI. If the current page is the landing page, we do not show the checkbox.

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx [hunk 5]
```diff
@@ -119,6 +125,9 @@ class PagesSidePane extends Component {
             }
             const currentLandingPage = this.props.dashboard.landingPage;
             this.props.dashboard.landingPage = newLandingPageId;
+            // Remove hidden flag from the new landing page
+            var page = this.getPageById(newLandingPageId);
+            delete page.hidden;
```
[the remark this answers is not in the document]
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: In the UI we do not prevent user to make a hidden page the landing page. That means, even though the page is marked as hidden, we show the option to make that the home page. What this logic does is, when the user makes a hidden page the landing page we remove its hidden property making it visible in the menu. Hope this is what you expect in your comment as well?

BTW, instead of deleting we can set `page.hidden = false;`.

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 6]
```diff
@@ -197,14 +197,35 @@ class DashboardViewPage extends Component {
         );
     }
 
+    /**
+     * Check if the given page is marked as hidden.
+     *
+     * @param {{*}} page Page object
+     * @returns {boolean} Is hidden
+     */
+    isPageHidden(page) {
```
[the remark this answers is not in the document]
Comment (lasanthaS) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Ack. Thanks!

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx (hunk 1 above)
Comment (lasanthaS) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Ack. Thanks!

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (this) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: > In the UI we do not prevent user to make a hidden page the landing page. That means, even though the page is marked as hidden, we show the option to make that the home page. What this logic does is, when the user makes a hidden page the landing page we remove its hidden property making it visible in the menu. Hope this is what you expect in your comment as well?

How about we
- disable the `Hide` checkbox if `Mark as home` is checked
- disable `Mark as home` checkbox if `Hide` is checked
This approch clearly convey that the landing page cannot be hidden and a hidden page cannot be the landing page.

>BTW, instead of deleting we can set page.hidden = false;
yes, lets do that.

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx [hunk 7]
```diff
@@ -170,6 +179,37 @@ class PagesSidePane extends Component {
...
+    /**
+     * Update dashboard page visibility.
+     *
+     * @param {string} pageId Page ID
+     * @param {boolean} hidden Is hidden
+     * @returns {Promise}
+     */
+    handlePageVisibility(pageId, hidden) {
+        return new Promise((resolve, reject) => {
+            const page = this.getPageById(pageId);
```
[the remark this answers is not in the document]
Comment (this) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: Ignore above.
Instead, set `page.hidden = false`  in `handleLandingPageUpdate` fucntion.

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (this) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: then lets remove :)

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (this) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: Why `delete page.hidden`?

IMO, instead of `delete` a property we should set an appropriate value to it. In this case, lets always set `page.hidden` to `true` or `false`. This will automatically migrate the dashboard JSON to new structure.

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: It's already there.

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx [hunk 8]
```diff
@@ -119,6 +125,9 @@ class PagesSidePane extends Component {
             }
             const currentLandingPage = this.props.dashboard.landingPage;
             this.props.dashboard.landingPage = newLandingPageId;
+            // Remove hidden flag from the new landing page
+            var page = this.getPageById(newLandingPageId);
```
Comment (this) [inline components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: instead of `var` use `const`
https://eslint.org/docs/rules/no-var

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx [hunk 9]
```diff
@@ -170,6 +179,33 @@ class PagesSidePane extends Component {
...
+     * @returns {Promise}
+     */
+    handlePageVisibility(pageId, hidden) {
+        return new Promise((resolve, reject) => {
+            const page = this.getPageById(pageId);
+            page.hidden = hidden;
+            this.props.updateDashboard()
+                .then(() => resolve())
+                .catch(() => {
+                    delete page.hidden;
```
Comment (this) [inline components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: Instead of deleting, we should revert the change.
See [`handleLandingPageUpdate` function](https://github.com/wso2/carbon-dashboards/pull/985/files#diff-817114f3567404d06ed20e6d48cbbd02R126).

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx [hunk 10]
```diff
@@ -197,16 +197,29 @@ class DashboardViewPage extends Component {
...
+     * @returns {XML} Html content
+     */
     renderPagesList() {
         const landingPage = this.dashboard.landingPage;
         const history = this.props.history;
-        return this.dashboard.pages.map((page) => {
+        let pagesList = [];
+        this.dashboard.pages.forEach((page) => {
+            if (page.hidden) {
+                console.log()
```
Comment (this) [inline components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: console.log

File: components/dashboards-web-component/src/designer/components/PageCard.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (this) [reply components/dashboards-web-component/src/designer/components/PageCard.jsx]: We should display the `Hide` checkbox even for the landing page. The checkbox needs to be disabled.

File: components/dashboards-web-component/src/designer/components/PageCard.jsx [hunk 11]
```diff
@@ -151,6 +151,18 @@ class PageCard extends Component {
...
+                                <Checkbox
+                                    label={<FormattedMessage id="designer.hide.page" defaultMessage="Hide Page" />}
+                                    defaultChecked={page.hidden}
+                                    onCheck={(e, checked) => {
```
Comment (this) [inline components/dashboards-web-component/src/designer/components/PageCard.jsx]: Add 
```js
disabled={ page.id === this.props.landingPageId }
```
so the checkbox will be disabled for the landing page.

File: components/dashboards-web-component/src/designer/components/PageCard.jsx [hunk 12]
```diff
@@ -151,6 +151,18 @@ class PageCard extends Component {
                                 disabled={page.id === this.props.landingPageId}
                                 onCheck={() => this.props.updateLandingPage(this.props.page.id)}
                             />
+                            {
+                                /* Show hide page option only for non-landing page pages. */
+                                page.id !== this.props.landingPageId &&
+                                <Checkbox
+                                    label={<FormattedMessage id="designer.hide.page" defaultMessage="Hide Page" />}
```
Comment (this) [inline components/dashboards-web-component/src/designer/components/PageCard.jsx]: Since this checkbox is in a page car, how about just `"Hide"`?

File: components/dashboards-web-component/src/designer/components/PageCard.jsx (hunk 11 above)
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PageCard.jsx]: What is the purpose of disabling and showing, when there is no purpose?

Comment (this) [reply components/dashboards-web-component/src/designer/components/PageCard.jsx]: Consistent UI. We shouldn't surprise the user with sudden UI changes (like appearing a new checkbox).
https://ux.stackexchange.com/a/112250/49172
https://ux.stackexchange.com/a/12761/49172

Page card for all pages should look same. Some pages can be special, like the landing page. In that case, some options might not be available such as `Hide`.

File: components/dashboards-web-component/src/designer/components/PageCard.jsx (hunk 12 above)
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PageCard.jsx]: Instead of "Hide",  better to have "Hide Page" since it provides some sense to the end-user about its functionality.

File: components/dashboards-web-component/src/viewer/DashboardViewPage.jsx (hunk 10 above)
Comment (lasanthaS) [reply components/dashboards-web-component/src/viewer/DashboardViewPage.jsx]: Removed

File: components/dashboards-web-component/src/designer/components/PagesSidePane.jsx (hunk 9 above)
Comment (lasanthaS) [reply components/dashboards-web-component/src/designer/components/PagesSidePane.jsx]: Ack

File: components/dashboards-web-component/src/designer/components/PageCard.jsx (hunk 12 above)
Comment (this) [reply components/dashboards-web-component/src/designer/components/PageCard.jsx]: Ok

File: components/dashboards-web-component/src/designer/components/PageCard.jsx [hunk 13]
```diff
@@ -151,6 +151,18 @@ class PageCard extends Component {
                                 disabled={page.id === this.props.landingPageId}
                                 onCheck={() => this.props.updateLandingPage(this.props.page.id)}
                             />
+                            {
+                                /* Enable hide page option only for non-landing page pages. */
+                                <Checkbox
+                                    label={<FormattedMessage id="designer.hide.page" defaultMessage="Hide Page" />}
+                                    defaultChecked={page.hidden}
+                                    disabled={page.id === this.props.landingPageId}
+                                    onCheck={(e, checked) => {
```
Comment (this) [inline components/dashboards-web-component/src/designer/components/PageCard.jsx]: With above, you will be able to change format this to a single line
```js
onCheck={(e, checked) => this.props.updatePageVisibility(this.props.page.id, checked)}
```
